File: test-suite/rust/buggy/header_injection.rs
use axum::extract::Query;
use std::collections::HashMap;

const CONTENT_DISPOSITION: &str = "Content-Disposition";

struct HeaderMap;
struct HeaderValue;
struct Request;
struct Response;
struct ResponseBuilder;

impl HeaderMap {
    fn get(&self, _name: &str) -> Option<HeaderValue> {
        Some(HeaderValue)
    }

    fn insert(&mut self, _name: &str, _value: impl AsRef<str>) {}
}

impl HeaderValue {
    fn to_str(&self) -> Result<&str, ()> {
        Ok("evil\r\nSet-Cookie: session=stolen")
    }
}

impl Request {
    fn headers(&self) -> HeaderMap {
        HeaderMap
    }
}

impl Response {
    fn builder() -> ResponseBuilder {
        ResponseBuilder
    }
}

impl ResponseBuilder {
    fn header(self, _name: &str, _value: impl AsRef<str>) -> Self {
        self
    }
}

fn query_value_in_header(Query(params): Query<HashMap<String, String>>) -> ResponseBuilder {
    let display_name = params.get("name").cloned().unwrap_or_default();
    Response::builder().header("X-Display-Name", display_name)
}

fn filename_in_content_disposition(Query(params): Query<HashMap<String, String>>) -> ResponseBuilder {
    let filename = params.get("filename").cloned().unwrap_or_default();
    Response::builder().header(CONTENT_DISPOSITION, format!("attachment; filename={filename}"))
}

fn request_header_reflected(req: Request) -> ResponseBuilder {
    let trace_id = req
        .headers()
        .get("x-trace-id")
        .and_then(|value| value.to_str().ok())
        .unwrap_or_default();
    Response::builder().header("X-Upstream-Trace", trace_id)
}

fn mutable_header_map(Query(params): Query<HashMap<String, String>>, mut response_headers: HeaderMap) {
    let mode = params.get("mode").cloned().unwrap_or_default();
    response_headers.insert("X-Mode", mode);
}

fn validate_after_setting(Query(params): Query<HashMap<String, String>>) -> Result<ResponseBuilder, &'static str> {
    let reason = params.get("reason").cloned().unwrap_or_default();
    let response = Response::builder().header("X-Return-Reason", &reason);
    if reason.contains('\r') || reason.contains('\n') {
        return Err("invalid header");
    }
    Ok(response)
}

File: test-suite/elixir/buggy/header_injection.ex
defmodule BuggyHeaderInjection do
  import Plug.Conn

  def display_name(conn) do
    name = conn.params["display_name"]

    conn
    |> put_resp_header("x-display-name", name)
    |> send_resp(200, "ok")
  end

  def download(conn) do
    filename = Map.get(conn.query_params, "filename")

    conn
    |> put_resp_header("content-disposition", "attachment; filename=\"#{filename}\"")
    |> send_resp(200, "ok")
  end

  def trace_header(conn) do
    trace =
      conn
      |> Plug.Conn.get_req_header("x-trace-id")
      |> List.first()

    Plug.Conn.put_resp_header(conn, "x-upstream-trace", trace)
  end

  def cookie_header(conn) do
    export_name = conn.cookies["export_name"]
    put_resp_header(conn, "x-export-name", export_name)
  end

  def content_type(conn) do
    media_type = conn.params["content_type"]
    put_resp_content_type(conn, media_type)
  end

  def mixed_location_and_display(conn) do
    tenant = conn.params["tenant"]
    put_resp_headers(conn, [{"location", "/safe"}, {"x-tenant", tenant}])
  end

  def validation_after_write(conn) do
    value = conn.params["customer"]
    response = put_resp_header(conn, "x-customer", value)

    if String.contains?(value, "\n") do
      raise ArgumentError, "invalid header"
    end

    response
  end
end
